### PR TITLE
#36044 Fix and enhance new autocompletion engine

### LINKED ISSUE #34251: Autocomplete on self joins does not work with "Enable experimental features" turned on
### Description

If "Enable experimental features" is turned on, autocomplete with self joins only works with alias mentioned last.

self_join_ac_select
self_join_ac_tri_select
self_join_no_props_select
self_join_no_props_tri_select

### DBeaver Version

Community Edition 24.1.0.202406021658

### Operating System

Windows 10

### Database and driver

MySQL 8.0.36-commercial
MySQL Connector/J mysql-connector-j-8.2.0 (Revision: 06a1f724497fd81c6a659131fda822c9e5085b6c)

### Steps to reproduce

1. Make sure "Enable experimental features" option is on
2. Write self join statement
3. Try autocomplete

### Additional context

_No response_

### LINKED ISSUE #35891: Sometimes autocomplete doesn't add number to the name of the table alias when join the same tables in one query
### Description

Alias name "f2" is expected here.

https://github.com/user-attachments/assets/d27a9381-0b1e-4c2b-b823-51c74e1729ab

### DBeaver Version

Enterprise Edition 24.2.3-202410032322

### Operating System

Windows 10

### Database and driver

PostgreSQL 16
PostgreSQL JDBC Driver 42.7.4

### Steps to reproduce

_No response_

### Additional context

_No response_

### LINKED ISSUE #35893: Autocomplete adds additional dot after column name in join condition when choosing one of suggestions
### Description

Use the query:
```
select * 
from film as f
join film_list as fl on fl.fid =
```

- press Tab to confirm the chosen item from suggestions

https://github.com/user-attachments/assets/1e251c8a-8222-4d5a-9e8e-87130b91f94a

### DBeaver Version

Enterprise Edition 24.2.3-202410032322

### Operating System

Windows 10

### Database and driver

PostgreSQL 16
PostgreSQL JDBC Driver 42.7.4

### Steps to reproduce

_No response_

### Additional context

_No response_

### LINKED ISSUE #36044: [New autocomplete] automatically pops up a join hint when a join is already typed
### Description

https://github.com/user-attachments/assets/f89b8299-69c9-47eb-92ff-b15fe49db514

### DBeaver Version

DBeaver 24.2.4

### Operating System

windows 10

### Database and driver

_No response_

### Steps to reproduce

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by E1izabeth**
- [x] https://github.com/dbeaver/dbeaver/pull/36434#issuecomment-2528477790
- [x] https://github.com/dbeaver/pro/issues/3818
- [x] Add autocompletion of join condition by non-compound foreign keys
       image
- [x] Fix completion right after the keyword
- [x] Fix qualified table names accomplishment
- [x] Fix keyword completion for some cases (now you can see more keywords in proposals than in the previous version)
- [x] Fix duplicated table aliases generation #35891
- [x] Fix additional dot inserting when applying the completion proposal in the join condition #35893
- [x] Autocomplete of tables columns in self-joins #34251

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.model.lsm/src/org/jkiss/dbeaver/model/stm/LSMInspections.java [hunk 1]
```diff
@@ -384,7 +367,7 @@ private static Collection<Transition> collectFollowingTerms(
         @NotNull ATNState initialState, Set<Integer> exceptRules,
         @NotNull Map<Integer, Boolean> reachabilityTests
     ) {
-        HashSet<ATNState> visited = new HashSet<>();
+        HashSet<Pair<ATNState, ListNode<Integer>>> visited = new HashSet<>();
```
Comment (HocKu7) [inline plugins/org.jkiss.dbeaver.model.lsm/src/org/jkiss/dbeaver/model/stm/LSMInspections.java:400]: Do you really need HeshSet rather than a simple Set?

**Review by HocKu7 — APPROVED** [review]

**Review by ShadelessFox — APPROVED** [review]

**Review by ShadelessFox — APPROVED** [review]

**Comment by MashaKorax** [comment]
if you put a dot after an alias and do autocomplete, the alias is duplicated


https://github.com/user-attachments/assets/cbba2083-1478-410a-9f17-5d453e4ec9dd

**Comment by MashaKorax** [comment]
left in LEFT JOIN recognized as alias

`select   from x left join y on x.x1 = y.y1 ;`


https://github.com/user-attachments/assets/ef6463f2-1014-4378-b058-ecf407b1b26c

**Comment by MashaKorax** [comment]
verified

**Pull request merged**